### PR TITLE
feat(rpc): optimize block validation with state caching

### LINKED ISSUE #12266: Use CachedReads for builder validations
### Describe the feature

We expect multiple request against the same state

https://github.com/paradigmxyz/reth/blob/a911104fe9ccb04c0946e0b11ac14337eec6b7e0/crates/rpc/rpc/src/validation.rs#L67-L67

https://github.com/paradigmxyz/reth/blob/a911104fe9ccb04c0946e0b11ac14337eec6b7e0/crates/rpc/rpc/src/validation.rs#L170-L171

the likelihood that multiple submissions access similar account/storage is likely high. We can use a pair of (Hash, CachedReads) that we clone before the request and merge after the the request, this additional overhead should be negligible compared to doing more disk io

see also:

https://github.com/paradigmxyz/reth/blob/a911104fe9ccb04c0946e0b11ac14337eec6b7e0/crates/revm/src/cached.rs#L32-L32
https://github.com/paradigmxyz/reth/blob/a911104fe9ccb04c0946e0b11ac14337eec6b7e0/crates/revm/src/cached.rs#L47-L47

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by lean-apple**
Closes https://github.com/paradigmxyz/reth/issues/12266. 

Introduces state caching with `CachedReads` for block validation.

Some next optimizations could be after checking the behavior  :
- Add cache size limits
- Implement cache cleanup strategy

### CODE REVIEW

File: crates/rpc/rpc/src/validation.rs [hunk 1]
```diff
@@ -78,6 +80,10 @@ pub struct ValidationApiInner<Provider: ChainSpecProvider, E> {
     executor_provider: E,
     /// Set of disallowed addresses
     disallow: HashSet<Address>,
+    /// Cached state reads to avoid redundant disk I/O across multiple validation attempts.
+    /// Maps block hash to cached reads for that block's state.
+    /// Uses async `RwLock` to safely handle concurrent validation requests.
+    cached_states: Arc<RwLock<HashMap<B256, CachedReads>>>,
```
Comment (mattsse) [inline crates/rpc/rpc/src/validation.rs:86]: I see, ValidationApiInner is clone, but it shouldn't be.
we can remove the clone impl and use Rwlock directly.

Comment (mattsse) [inline crates/rpc/rpc/src/validation.rs:86]: we only really need one instance here, then we don't need to do manual cleanup.
validations only target the head state so we only need to store `(B256, CachedReads)`

**Review by mattsse — CHANGES_REQUESTED** [review]
sorry for the late review.

this looks great, only have two nits really

File: crates/rpc/rpc/src/validation.rs (hunk 1 above)
Comment (lean-apple) [reply crates/rpc/rpc/src/validation.rs]: Ok I try to remove a couple of `Clone`

Comment (lean-apple) [reply crates/rpc/rpc/src/validation.rs]: Got it, I 've removed the `HashMap`

**Review by mattsse — APPROVED** [review]
great!

I only extracted the cached_read ops into standalone functions

**Comment by lean-apple** [comment]
Sorry though it was already merged

**Pull request merged**